float8_experimental/float8_dynamic_linear.py

```diff
@@ -95,7 +113,110 @@ def from_float(cls, mod, emulate: bool = False) -> "Float8DynamicLinear":
                 "bias": False,
             }
             new_mod = cls(**super_kwargs)
-        new_mod.weight = mod.weight
+        new_mod.weight = (
+            nn.Parameter(Float8DynamicLinearWeightTensor(mod.weight, emulate))
+            if config.enable_fsdp_fp8_all_gather
+            else mod.weight
+        )
         new_mod.bias = mod.bias
         new_mod.emulate = emulate
         return new_mod
+
+
+# FSDP pads its local tensor on dim-0. The subclass should be preserved such
+# that the padded local tensor (and any transformations like copying to GPU)
+# is of the subclass as well.
+_ops_to_preserve_subclass = {
+    torch.ops.aten.empty_like.default,
+    torch.ops.aten.new_zeros.default,
+    torch.ops.aten.slice.Tensor,
+    torch.ops.aten.copy_.default,
+    torch.ops.aten.view.default,
+    torch.ops.aten.as_strided.default,
+    torch.ops.aten._to_copy.default,
+    torch.ops.aten._pin_memory.default,
+}
+
+
+class Float8DynamicLinearWeightTensor(torch.Tensor):
```

Comment:
maybe something like `TensorWithFloat8CastDefined` could be a more descriptive name?

`Float8Tensor` currently means "tensor with raw data in float8 precision, mimicing a high precision tensor

this object seems to be pretty different, would be good if the name is also different.

Reply:
I think the name may want to mention that the cast is using dynamic scaling (since delayed scaling would use a different cast and require a slightly different implementation of the pre-all-gather transformation).

Maybe `TensorWithDynamicFloat8Cast`? I also noticed that all of the existing subclasses are `<...>Tensor` (i.e. have a `Tensor` as suffix) 🤔 .

Reply:
`Float8AwareTensor`? `DynamicallyCastableToFloat8Tensor`?  I'm open, I think I just care that the name emphasizes that this is not a tensor stored in float8 precision, this is a high precision tensor that knows how to cast itself to float8.

Reply:
I renamed it to `WeightWithDynamicFloat8CastTensor` for now since we may (eventually) want a subclass for delayed scaling too, in which case we could perhaps do `WeightWithDelayedFloat8CastTensor`.

(The reason why a separate subclass might be easier than trying to reuse the same one is that the cast function is somewhat different with different args, so unifying it might be more clunky than beneficial.)